src/vs/workbench/contrib/inlineChat/electron-sandbox/inlineChatQuickVoice.ts

```diff
@@ -0,0 +1,226 @@
+	private readonly _elements = h('.inline-chat-quick-voice@main', [
+		h('span@mic'),
+		h('span@message'),
+	]);
+
+	constructor(private readonly _editor: ICodeEditor) {
+		this._domNode.appendChild(this._elements.root);
+		this._domNode.style.zIndex = '1000';
+		reset(this._elements.mic, renderIcon(Codicon.micFilled));
+	}
+
+	getId(): string {
+		return 'inlineChatQuickVoice';
+	}
+
+	getDomNode(): HTMLElement {
+		return this._domNode;
+	}
+
+	getPosition(): IContentWidgetPosition | null {
+		if (!this._editor.hasModel()) {
+			return null;
+		}
+		const selection = this._editor.getSelection();
+		// const position = this._editor.getPosition();
+		return {
+			position: selection.getPosition(),
+			preference: [
+				selection.getPosition().equals(selection.getStartPosition()) ? ContentWidgetPositionPreference.ABOVE : ContentWidgetPositionPreference.BELOW,
+				ContentWidgetPositionPreference.EXACT
+			]
+		};
+	}
+
+	beforeRender(): IDimension | null {
+		const lineHeight = this._editor.getOption(EditorOption.lineHeight);
+		this._elements.main.style.lineHeight = `${lineHeight}px`;
+		this._elements.main.style.height = `${lineHeight}px`;
+		return null;
+	}
+
+	// ---
+
+	updateInput(input: string | undefined): void {
+		this._elements.message.textContent = input ?? '';
+	}
+
+	active(): void {
+		this._elements.main.classList.add('recording');
+	}
+
+	reset(): void {
+		this._elements.main.classList.remove('recording');
+		this.updateInput(undefined);
+	}
+}
+
+export class InlineChatQuickVoice implements IEditorContribution {
+
+	static readonly ID = 'inlineChatQuickVoice';
+
+	static get(editor: ICodeEditor): InlineChatQuickVoice | null {
+		return editor.getContribution<InlineChatQuickVoice>(InlineChatQuickVoice.ID);
+	}
+
+	private readonly _ctxQuickChatInProgress: IContextKey<boolean>;
+	private readonly _widget: QuickVoiceWidget;
+	private _finishCallback?: (abort: boolean) => void;
+
+	constructor(
+		private readonly _editor: ICodeEditor,
+		@ISpeechService private readonly _speechService: ISpeechService,
+		@IContextKeyService contextKeyService: IContextKeyService,
+	) {
+		this._widget = new QuickVoiceWidget(this._editor);
+		this._ctxQuickChatInProgress = CTX_QUICK_CHAT_IN_PROGRESS.bindTo(contextKeyService);
+	}
+
+	dispose(): void {
+		this._finishCallback?.(true);
+	}
+
+	start() {
+
+		const cts = new CancellationTokenSource();
+		this._editor.addContentWidget(this._widget);
+		this._ctxQuickChatInProgress.set(true);
+
+		let message: string | undefined;
+
+		const session = this._speechService.createSpeechToTextSession(cts.token);
+		const listener = session.onDidChange(e => {
+
+			if (cts.token.isCancellationRequested) {
+				return;
+			}
+
+			switch (e.status) {
+				case SpeechToTextStatus.Started:
+					this._widget.active();
+					break;
+				case SpeechToTextStatus.Stopped:
+					break;
+				case SpeechToTextStatus.Recognizing:
+				case SpeechToTextStatus.Recognized:
+					message = e.text;
```

Comment:
Fyi there is a slight difference between the two:
- recognizing: will spit out words until pause
- recognized: will spit out the entire sequence of words from start or last pause to pause

I think the idea from the SDK is that a UI would show the former with a grey background (sort of unconfirmed) and the latter as final confirmed result. From my testing I have not seen a big difference in terms of quality, but here is how I apply it in my code:

https://github.com/microsoft/vscode/blob/86d1a94cca007f64e442c1aa0b6af60a40c1ceea/src/vs/workbench/contrib/chat/electron-sandbox/actions/voiceChatActions.ts#L286-L302

In other words: I append `Recognizing` text with whitespace until I receive `Recognized` where I drop all `Recognizing` text and append the result of `Recognized`